### PR TITLE
feat: add upgrade access button to user settings page

### LINKED ISSUE #1394: Feature: Add upgrade access button to user settings page
### Type of feature

🍕 Feature

### Current behavior

This would add a button in the user settings page to upgrade the user to get go through the checkout process and get access to reports. This is currently only available on the `Reports` tab for logged in users. 

### Suggested solution

Add the button under the email preferences.

image

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by babblebey**
## Description

This Pull Request adds the `StripeCheckoutButton` component to the User Settings page.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Fixes #1394 

## Mobile & Desktop Screenshots/Recordings

[screencast-localhost_3000-2023.07.20-20_43_11.webm](https://github.com/open-sauced/insights/assets/25631971/28168090-37f5-4129-94d1-0386bc4a4334)

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

NA

### CODE REVIEW

File: components/organisms/UserSettingsPage/user-settings-page.tsx [hunk 1]
```diff
@@ -371,6 +372,12 @@ const UserSettingsPage = ({ user }: userSettingsPageProps) => {
               Update Preferences
             </Button>
           </div>
+          <div className="flex flex-col gap-3 order-first md:order-last">
+            <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
+            <div className="text-center p-3 border border-dashed rounded-xl border-light-slate-8">
```
Comment (brandonroberts) [inline components/organisms/UserSettingsPage/user-settings-page.tsx:377]: I don't think it needs a border around the box, and if the button can be left aligned to be consistent with the other ones.

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: components/organisms/UserSettingsPage/user-settings-page.tsx (hunk 1 above)
Comment (babblebey) [reply components/organisms/UserSettingsPage/user-settings-page.tsx:377]: Sure.

I will go on and refactor the `StripeCheckoutButton` component to be more customisable for usage in this context. Its current set-up is hard coded to work this way as integrated with the Dashboard's `Reports` component.

Comment (bdougie) [reply components/organisms/UserSettingsPage/user-settings-page.tsx:377]: Making the component more flexible would be ideal in this case.

Comment (babblebey) [reply components/organisms/UserSettingsPage/user-settings-page.tsx]: Done 🍕

File: components/organisms/StripeCheckoutButton/stripe-checkout-button.tsx [hunk 2]
```diff
@@ -32,17 +34,11 @@ const StripeCheckoutButton = () => {
...
-        <form onSubmit={handleFormSubmit}>
-          <Button variant="primary" className="w-52 h-[38px] flex justify-center">
-            Upgrade Access
-          </Button>
-        </form>
-      </p>
-    </div>
+    <form onSubmit={handleFormSubmit}>
+      <Button {...props}>
+        { children || "Upgrade Access" }
```
Comment (NsdHSO) [inline components/organisms/StripeCheckoutButton/stripe-checkout-button.tsx:39]:
```suggestion
        {children || "Upgrade Access"}
```

File: components/organisms/UserSettingsPage/user-settings-page.tsx [hunk 3]
```diff
@@ -371,6 +373,15 @@ const UserSettingsPage = ({ user }: userSettingsPageProps) => {
               Update Preferences
             </Button>
           </div>
+          <div className="flex flex-col gap-6 order-first md:order-last">
+            <div className="flex flex-col gap-3">
+              <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
+              <div className="w-full sm:max-w-[330px]">
```
Comment (NsdHSO) [inline components/organisms/UserSettingsPage/user-settings-page.tsx:379]:
```suggestion
              <div className="w-full sm:max-w-80]">
```

**Comment by babblebey** [comment]
Changes Made to make the `StripeCheckoutButton` component more flexible:

- Modified the `StripeCheckoutButton` component to accept `children` and `props` of type `StripeCheckoutButtonProps`. The `children` prop allows users to customize the text content of the button, while the `props` can be used to pass additional styling or behavior options.
- Removed the unnecessary `<div>` element and replaced it with a `<form>` element to wrap the `Button`. This ensures better semantic structure for form submission.
- Removed the redundant `<p>` element around the `<form>` element, as it is no longer necessary.
- Updated the `Button` component usage to spread the `props` object so that any additional props provided to `StripeCheckoutButton` can be passed down to the `Button` component.

Overall, these changes have made the `StripeCheckoutButton` component more flexible and reusable, allowing users to easily customize the button's appearance and behavior based on their specific use cases. _Sample Use Cases seen below_

**The dashboard `Reports` Component use case**

```javascript
<>
  <Text>Upgrade to a subscription to gain access to generate custom reports!</Text>
  <p className="flex justify-center py-4 px-2">
    // HERE
    <StripeCheckoutButton variant="primary" className="w-52 h-[38px] flex justify-center"/>
  </p>
</>
```

screenshot-localhost_3000-2023 07 21-11_30_31

**The User Settings page use case**

```javascript
<div className="flex flex-col gap-6 order-first md:order-last">
  <div className="flex flex-col gap-3">
    <label className="text-2xl font-normal text-light-slate-11">Upgrade Access</label>
    <div className="w-full sm:max-w-[330px]">
      <Text>Upgrade to a subscription to gain access to generate custom reports!</Text>
    </div>
  </div>
  // HERE
  <StripeCheckoutButton variant="primary" />
</div>
```

screenshot-localhost_3000-2023 07 21-12_50_04

**The User Settings page use case (with custom button content)**
```javascript
<StripeCheckoutButton variant="primary">
  Subscribe Now
</StripeCheckoutButton>
```

screenshot-localhost_3000-2023 07 21-12_51_39

**Comment by brandonroberts** [comment]
@babblebey changes look better. Got some build/test failures on the PR preview

**Comment by babblebey** [comment]
> changes look better. Got some build/test failures on the PR preview

I was out of sync with the new build configuration #1396

Now Fixed 🍕

**Comment by brandonroberts** [comment]
@babblebey thanks. One more addition is that button shouldn't be displayed if the user account already has access. The `hasReports` property from the `useSession` hook gives you this info https://github.com/open-sauced/insights/blob/beta/lib/hooks/useSession.ts#L43

Should be good to go then

**Comment by babblebey** [comment]
> thanks. One more addition is that button shouldn't be displayed if the user account already has access. 

Haha, funny how I didn't think about that logic 

Will do 🍕

**Review by brandonroberts — APPROVED** [review]

**Pull request merged**